File: apps/legacy-cli/modules/state.py
#!./venv/bin/python3
import yaml
from modules.logs import LOGS

myLOGS = LOGS()

class IDPSTATE:
    # Define sensitive variables to be filtered out
    sensitive_vars = ["applications_argo_access_token", "gitops_argo_access_token"]

    def save_infra_facts(self, config, outputs=None, dry_run=False):
        myLOGS.log("normal", "Saving infra-facts.yaml")
        to_yaml = {}

        # Include top-level variables from the config, but filter out sensitive variables
        myLOGS.log("normal", "Including top-level variables from the config.")
        for key, value in vars(config).items():
            if key not in self.sensitive_vars and isinstance(value, (float, str, int, bool)):
                to_yaml[key] = value

        # Include terraform outputs
        if outputs:
            myLOGS.log("normal", "Including terraform outputs.")
            for key, output in outputs.items():
                to_yaml[key] = output["value"]
        else:
            if dry_run:
                myLOGS.log("normal", "No Terraform outputs found. This is expected in dry-run mode. "
                                        "The infra-facts.yaml will not be fully populated with outputs...")
            else:
                myLOGS.log("critical", "No Terraform outputs found in non-dry-run mode. Exiting.")
                exit(1)

        # Save the infra-facts YAML
        output = "temp/infra-facts.yaml"
        myLOGS.log("normal", f'Saving infra-facts.yaml file to: {output}')
        yaml_string = yaml.dump({"infra-services": to_yaml})
        with open(output, "w") as text_file:
            text_file.write(yaml_string)

File: packages/core/state/state.go
// SPDX-FileCopyrightText: 2026 Alethia Labs <legal@alethialabs.io>
// SPDX-License-Identifier: AGPL-3.0-only

package state

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"

	"github.com/alethialabs-io/alethialabs/packages/core/utils"
	"gopkg.in/yaml.v3"
)

var sensitiveFields = map[string]bool{
	"applications_argo_access_token": true,
	"gitops_argo_access_token":       true,
}

type State struct{}

func NewState() *State {
	return &State{}
}

func (s *State) SaveInfraFacts(rawConfig map[string]interface{}, outputs map[string]interface{}, dryRun bool, logger *utils.Logger) error {
	logger.Info("Saving infra-facts.yaml", "state")
	toYAML := make(map[string]interface{})

	for k, v := range rawConfig {
		if sensitiveFields[k] {
			continue
		}
		switch v.(type) {
		case float64, string, int, bool:
			toYAML[k] = v
		}
	}

	if len(outputs) > 0 {
		logger.Info("Including OpenTofu outputs.", "state")
		for key, output := range outputs {
			toYAML[key] = output
		}
	} else {
		if dryRun {
			logger.Warn("No OpenTofu outputs found. This is expected in dry-run mode.", "state")
		} else {
			return fmt.Errorf("no OpenTofu outputs found in non-dry-run mode")
		}
	}

	finalMap := map[string]interface{}{
		"infra-services": toYAML,
	}

	yamlData, err := yaml.Marshal(finalMap)
	if err != nil {
		return fmt.Errorf("failed to marshal state to YAML: %w", err)
	}

	tempDir := "temp"
	if err := os.MkdirAll(tempDir, 0755); err != nil {
		return fmt.Errorf("failed to create temp directory: %w", err)
	}

	outputPath := filepath.Join(tempDir, "infra-facts.yaml")
	if err := os.WriteFile(outputPath, yamlData, 0644); err != nil {
		return fmt.Errorf("failed to write infra-facts.yaml: %w", err)
	}

	logger.Info(fmt.Sprintf("Saved infra-facts.yaml to %s", outputPath), "state")
	return nil
}

func RawConfigFromFullConfig(fullConfig *string) (map[string]interface{}, error) {
	if fullConfig == nil || *fullConfig == "" {
		return nil, fmt.Errorf("no raw config available")
	}
	var raw map[string]interface{}
	if err := json.Unmarshal([]byte(*fullConfig), &raw); err != nil {
		return nil, fmt.Errorf("failed to parse full config: %w", err)
	}
	return raw, nil
}
